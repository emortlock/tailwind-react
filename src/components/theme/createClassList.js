import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
