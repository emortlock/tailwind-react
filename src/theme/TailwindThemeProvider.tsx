import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
