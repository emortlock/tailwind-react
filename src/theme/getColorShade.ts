const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
