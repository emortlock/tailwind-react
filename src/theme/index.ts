export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
