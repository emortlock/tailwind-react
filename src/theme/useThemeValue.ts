import { tailwindPropToClassName } from 'tailwind-react-primitives'

export const useThemeValue = (
  prefix: string,
  value: unknown,
  userClassNames = '',
) => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
