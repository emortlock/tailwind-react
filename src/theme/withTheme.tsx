import type { Theme } from './TailwindTheme'

import React from 'react'

import { TailwindTheme } from './TailwindTheme'

export interface WithTheme {
  theme: Theme
}

export const withTheme = <P extends React.PropsWithChildren<{ theme: Theme }>>(
  Component: React.FC<P>,
) => {
  const WithTheme: React.FC<Omit<P, 'theme'>> = (props) => (
    <TailwindTheme.Consumer>
      {(theme) => {
        // @ts-ignore FIXME: Constraint error
        return <Component {...props} theme={theme} />
      }}
    </TailwindTheme.Consumer>
  )

  WithTheme.displayName = `WithTheme(${Component.displayName})`

  return WithTheme
}
