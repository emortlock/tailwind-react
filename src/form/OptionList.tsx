import type { ListProps } from '../list'
import type { WithTheme } from '../theme'
import type { WithField } from './Field'

import React from 'react'

import { RawList as List } from '../list'
import { withTheme } from '../theme'

export interface OptionListProps<E extends HTMLElement = HTMLUListElement>
  extends WithTheme,
    WithField,
    ListProps<E> {
  name: string
  checkbox?: boolean
  invalid?: boolean
}

export const RawOptionList = <E extends HTMLElement = HTMLUListElement>({
  theme,
  children,
  name,
  checkbox,
  field = {},
  invalid,
  ...rest
}: OptionListProps<E>) => {
  const [checked, setChecked] = React.useState<string[]>([])

  const handleChange = React.useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const { value } = e.target

      if (checkbox) {
        return setChecked(
          checked.includes(value)
            ? checked.filter((checkedVal) => checkedVal !== value)
            : [...checked, value],
        )
      }

      return setChecked([value])
    },
    [checked],
  )

  const describedBy = [field.labelId, field.errorId, field.helpId].filter(
    Boolean,
  )
  const isInvalid = field.invalid || invalid

  return (
    <List<E> theme={theme} m={{ b: 0 }} {...rest}>
      {React.Children.map(children, (child) => {
        if (!child || !React.isValidElement(child)) return child

        const value = child.props && child.props.value

        return React.cloneElement(child, {
          'id': `${field.inputId}-${value}`,
          'name': checkbox ? `${name}[]` : name,
          'checked': checked.includes(value),
          'onChange': handleChange,
          'aria-invalid': isInvalid || undefined,
          'aria-describedby': describedBy.length
            ? describedBy.join(' ')
            : undefined,
        })
      })}
    </List>
  )
}

export const OptionList = withTheme<OptionListProps>(RawOptionList)
