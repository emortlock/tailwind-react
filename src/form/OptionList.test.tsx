import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawOptionList as OptionList } from './OptionList'
import { defaultTheme } from '../theme'

import { Radio } from '.'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign({ theme: defaultTheme, name: 'test' }, testProps)

  const wrapper = render(
    <OptionList {...props}>
      <Radio name="option-input" value="yes" label="Yes" />
      <Radio name="option-input" value="no" label="No" />
      <Radio name="option-input" value="maybe" label="Maybe" />
    </OptionList>,
  )

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('OptionList', () => {
  it('renders matching snapshot', () => {
    const [{ asFragment }] = setup()

    expect(asFragment()).toMatchSnapshot()
  })
})
