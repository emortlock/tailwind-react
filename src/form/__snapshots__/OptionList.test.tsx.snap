// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionList renders matching snapshot 1`] = `
<DocumentFragment>
  <ul
    class="mb-0"
  >
    <li
      class="mb mr"
    >
      <label
        class="font-bold leading-normal inline-block mb-2 select-none cursor-pointer flex items-center focus:outline-none focus:shadow-outline"
        role="button"
        tabindex="0"
      >
        <input
          class="visually-hidden"
          id="undefined-yes"
          name="test"
          tabindex="-1"
          type="radio"
          value="yes"
        />
        <span
          class="inline-block flex items-center justify-center rounded-full h-4 w-4 border mr-2"
        />
        <span
          class="inline-block leading-tight font-normal"
        >
          Yes
        </span>
      </label>
    </li>
    <li
      class="mb mr"
    >
      <label
        class="font-bold leading-normal inline-block mb-2 select-none cursor-pointer flex items-center focus:outline-none focus:shadow-outline"
        role="button"
        tabindex="0"
      >
        <input
          class="visually-hidden"
          id="undefined-no"
          name="test"
          tabindex="-1"
          type="radio"
          value="no"
        />
        <span
          class="inline-block flex items-center justify-center rounded-full h-4 w-4 border mr-2"
        />
        <span
          class="inline-block leading-tight font-normal"
        >
          No
        </span>
      </label>
    </li>
    <li
      class="mb mr"
    >
      <label
        class="font-bold leading-normal inline-block mb-2 select-none cursor-pointer flex items-center focus:outline-none focus:shadow-outline"
        role="button"
        tabindex="0"
      >
        <input
          class="visually-hidden"
          id="undefined-maybe"
          name="test"
          tabindex="-1"
          type="radio"
          value="maybe"
        />
        <span
          class="inline-block flex items-center justify-center rounded-full h-4 w-4 border mr-2"
        />
        <span
          class="inline-block leading-tight font-normal"
        >
          Maybe
        </span>
      </label>
    </li>
  </ul>
</DocumentFragment>
`;
